### PR TITLE
[integrations] [microsft message trace] - Deprecated basic auth and added support for Oauth2

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Enhancement

## What does this PR do?
This deprecates the basic auth fields and adds support for OAuth2. This had to be done because Microsoft has 
completely ended support of basic auth in Message Trace APIs since 31st March 2023. This has been also mentioned
here: https://techcommunity.microsoft.com/t5/exchange-team-blog/basic-authentication-deprecation-in-exchange-online-time-s-up/ba-p/3695312

This change should also fix the SDH: https://github.com/elastic/sdh-beats/issues/3262

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Relates [Microsoft Exchange Online Message Trace integration Basic auth deprecated](https://github.com/elastic/sdh-beats/issues/3262)

### CODE REVIEW

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md [hunk 1]
```diff
@@ -15,6 +15,25 @@ The `log` dataset collects the Microsoft Exchange Online Message Trace logs.
 
 [Log Documentation](https://docs.microsoft.com/en-us/previous-versions/office/developer/o365-enterprise-developers/jj984335(v=office.15))
 
+## Configuring with OAuth2
+The basic auth fields have been deprecated from this integration since Microsoft has deprecated them and they no longer work.
+Going forward OAuth2 needs to be configured for working with Microsoft Exchange Online Message Trace APIs.
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:20]:
```suggestion
The basic authentication fields have been deprecated from this integration as Microsoft has deprecated and disabled basic authentication for Exchange Online. See the [deprecation notification](https://learn.microsoft.com/en-us/exchange/clients-and-mobile-in-exchange-online/deprecation-of-basic-authentication-exchange-online) for details.
In order to continue using the Microsoft Exchange Online Message Trace you will need to enable and configure OAuth2 authentication.
```

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md [hunk 2]
```diff
@@ -15,6 +15,25 @@ The `log` dataset collects the Microsoft Exchange Online Message Trace logs.
...
+2) Click on "Azure Active Directory" in the left-hand menu.
+3) Select "App registrations" and click "New registration".
+4) Enter a name for your application, select "Accounts in this organizational directory only" for "Supported account types", and enter the redirect 
+   URI for your application.
+5) Click "Register" to create the application.
+6) On the application page, make note of the "Application (client) ID" (which is your client ID) and the "Directory (tenant) ID" (which is your 
+   tenant ID).
+7) Under "Certificates & secrets", click "New client secret" to create a new secret. Make note of the secret value (which is your client secret).
+
+With these credentials in hand, you can now configure the integration with the appropriate parameters.
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:36]:
```suggestion
With these credentials in hand, you can now configure the integration with the appropriate parameters.

```

File: packages/microsoft_exchange_online_message_trace/manifest.yml [hunk 3]
```diff
@@ -27,31 +27,82 @@ policy_templates:
...
             type: text
-            title: Username
-            description: Username to access Exchange Online
+            title: OAuth Server URL
+            multi: false
             required: true
-          - name: password
-            type: password
-            title: Password
-            description: Password to access Exchange Online
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/manifest.yml:53]: This is more a removal than a deprecation. Perhaps reword the documentation.

**Review by efd6 — COMMENTED** [review]
Does the documentation for the log collection by PowerShell need to be updated?

**Comment by ShourieG** [comment]
@efd6 have updated the docs along with the powershell script.

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md [hunk 4]
```diff
@@ -64,22 +75,23 @@ $pageSize = "1000"
 # This would then be ingested via the integration
 $output_location = "C:\temp\messageTrace.json"
 
-$password = ConvertTo-SecureString $password -AsPlainText -Force
-$Credential = New-Object System.Management.Automation.PSCredential ($username, $password)
+$token = Get-ExoOAuthAccessToken -ClientId $clientId -ClientSecret $clientSecret -TenantId $tenantId -Scopes $scopes
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:78]: Is there a link for the docs for this cmdlet? I can't find it and it doesn't appear to be in the module.

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md [hunk 5]
```diff
@@ -64,22 +75,23 @@ $pageSize = "1000"
...
 $endDate = Get-Date
 
-Connect-ExchangeOnline -Credential $Credential
 $paginate = 1
 $page = 1
 $output = @()
 while ($paginate -eq 1)
 {
-    $messageTrace = Get-MessageTrace -PageSize $pageSize -StartDate $startDate -EndDate $endDate -Page $page
+    $messageTrace = Get-MessageTrace -PageSize $pageSize -StartDate $startDate -EndDate $endDate -Page $page -AccessToken $token.AccessToken
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:90]: This doesn't look right; [AFAICS](https://learn.microsoft.com/en-us/powershell/module/exchange/get-messagetrace?view=exchange-ps) `Get-MessageTrace` doesn't accept `-AccessToken`.

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md [hunk 6]
```diff
@@ -33,18 +33,27 @@ You can follow these steps to create an Azure AD application:
...
 The following sample Powershell script may be used to get the logs and put them into a JSON file that can then be
 consumed by the logfile input:
 
 Prerequisites:
 
+Install the Exchange Online Management module by running the following command: 
+
 ````powershell
 Install-Module -Name ExchangeOnlineManagement
 ````
```
Comment (efd6) [inline packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:49]: Do we also need to do this and the import for `ExchangePowerShell` or do we assume that?

Comment (ShourieG) [reply packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:49]: We are currently assuming this

Comment (efd6) [reply packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:49]: Fair enough.

File: packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md (hunk 4 above)
Comment (ShourieG) [reply packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md]: The modified script has been removed and a disclaimer added to the old one.

**Comment by ShourieG** [comment]
@efd6 the doc has been updated with a Disclamer

**Pull request merged**